testsuite/tests/typing-unique/unique_analysis.ml

```diff
@@ -0,0 +1,465 @@
+[%%expect{|
+val unique_id : unique_ 'a -> unique_ 'a = <fun>
+|}]
+
+let shared_id : 'a -> 'a = fun x -> x
+[%%expect{|
+val shared_id : 'a -> 'a = <fun>
+|}]
+
+type box = { x : int }
+[%%expect{|
+type box = { x : int; }
+|}]
+
+let update : unique_ box -> unique_ box = unique_id
+[%%expect{|
+val update : unique_ box -> unique_ box = <fun>
+|}]
+
+
+(* testing Texp_ifthenelse  *)
+
+let branching (unique_ x : float) = unique_ if true then x else x
+[%%expect{|
+val branching : unique_ float -> float = <fun>
+|}]
+
+(* whether we constrain uniqueness or linearity is irrelavant
+   for testing uniqueness analysis. Therefore, in the rest we
+   will only constrain uniqueness *)
+let branching (once_ x : float) = if true then x else x
+[%%expect{|
+val branching : once_ float -> once_ float = <fun>
+|}]
+
+let branching b =
+  let unique_ r = { x = 23 } in
+  if b then update r
+       else update r
+[%%expect{|
+val branching : bool -> box = <fun>
+|}]
+
+let sequence (unique_ x : float) = unique_ let y = x in (x, y)
+[%%expect{|
+Line 1, characters 57-58:
+1 | let sequence (unique_ x : float) = unique_ let y = x in (x, y)
+                                                             ^
+Error: This is used uniquely here so cannot be used twice.  Another use is
+Line 1, characters 60-61:
+1 | let sequence (unique_ x : float) = unique_ let y = x in (x, y)
+                                                                ^
+
+|}]
+
+let sequence =
+  let r = { x = 23 } in
+  let s = update r in
+  let t = update s in
+  t
+[%%expect{|
+val sequence : box = {x = 23}
+|}]
+
+let sequence =
+  let r = { x = 23 } in
+  let _s = update r in
+  let t = update r in
+  t
+[%%expect{|
+Line 3, characters 18-19:
+3 |   let _s = update r in
+                      ^
+Error: This is used uniquely here so cannot be used twice.  Another use is
+Line 4, characters 17-18:
+4 |   let t = update r in
+                     ^
+
+|}]
+
+let children_unique (unique_ xs : float list) = match xs with
+  | [] -> (0., [])
+  | x :: xx -> unique_ (x, xx)
+[%%expect{|
+val children_unique : unique_ float list -> float * float list = <fun>
+|}]
+
+let borrow_match (unique_ fs : 'a list) = unique_ match fs with
+  | [] -> []
+  | x :: xs as gs -> gs
+[%%expect{|
+val borrow_match : unique_ 'a list -> 'a list = <fun>
+|}]
+
+let borrow_match (unique_ fs : 'a list) = unique_ match fs with
+    | [] -> []
+    | x :: xs -> fs
+[%%expect{|
+val borrow_match : unique_ 'a list -> 'a list = <fun>
+|}]
+
+let dup_child (unique_ fs : 'a list) = unique_ match fs with
+  | [] -> ([], [])
+  | x :: xs as gs -> (gs, xs)
+[%%expect{|
+Line 3, characters 22-24:
+3 |   | x :: xs as gs -> (gs, xs)
+                          ^^
+Error: This is used uniquely here so cannot be used twice.  Another use is
+Line 3, characters 26-28:
+3 |   | x :: xs as gs -> (gs, xs)
+                              ^^
+  The latter is the position-1 element of the constructor "::" in the value which is the former
```

Comment:
Yeah I think for simple cases (e.g. direct parent-child) it's better to keep it simple; but on the other hand, I'm afraid that in the complex case of multi-level relation the user would want to know the exact relation between the parent and the child.

Reply:
We discussed last time that I could try to print the path in ocaml syntax:
> Error: Cannot use `xs` here as it has already been used uniquely as part of `gs`, where `gs = (xs, _)`

But in retrospect this would add some extra complexity (for example, the projection need to include the # of fields in the tuple) so I think it's not worth it. I will just do the simple stuff.